source/common/upstream/cluster_manager_impl.cc

```diff
@@ -835,24 +835,35 @@ ClusterManagerImpl::loadCluster(const envoy::config::cluster::v3::Cluster& clust
   auto& lb = new_cluster_pair_or_error->second;
   Cluster& cluster_reference = *new_cluster;
 
+  const auto cluster_info = cluster_reference.info();
+
   if (!added_via_api) {
-    if (cluster_map.find(new_cluster->info()->name()) != cluster_map.end()) {
+    if (cluster_map.find(cluster_info->name()) != cluster_map.end()) {
       throw EnvoyException(
-          fmt::format("cluster manager: duplicate cluster '{}'", new_cluster->info()->name()));
+          fmt::format("cluster manager: duplicate cluster '{}'", cluster_info->name()));
     }
   }
 
-  if (cluster_reference.info()->lbType() == LoadBalancerType::ClusterProvided && lb == nullptr) {
+  // Check if the cluster provided load balancing policy is used. We need handle it as special
+  // case.
+  bool cluster_provided_lb = cluster_info->lbType() == LoadBalancerType::ClusterProvided;
+  if (cluster_info->lbType() == LoadBalancerType::LoadBalancingPolicyConfig) {
```

Comment:
nit: the following code does the following:
```
bool flag = <condition1>;
if (<condition1>) {
  flag = <condition2>;
}
```
which is somewhat confusing.

Taking a step back, what should happen if `cluster_info->lbType() == LoadBalancerType::ClusterProvided`, but the `typed_lb_factory->name()` is not `"envoy.load_balancing_policies.cluster_provided"`?
I'm assuming it is an invalid configuration, and so it should probably be rejected.
If you add a rejection part, then `bool cluster_provided_lb = ...` can be converted to `const bool cluster_provided_lb = ...` and no need to override the value.

The above assumes that both `lbType()` and `loadBalancerFactory()` settings are mandatory. If not, then the current change "breaks backwards compatibility" when the `lbType()` is `ClusterProvided` but the provided factory is different.

Reply:
I think the logic accually is:
```
bool flag = <condition1>;
if (<condition2>) {
  flag = <condition3>;
}
```

For now, we will always has lbType(). But the `loadBalancerFactory()` is set iff the `lbType() == LoadBalancerType::LoadBalancingPolicyConfig`.

So, the following case will neve happen. We will only check the name when `lbType() == LoadBalancerType::LoadBalancingPolicyConfig`.
>  if cluster_info->lbType() == LoadBalancerType::ClusterProvided, but the typed_lb_factory->name() is not "envoy.load_balancing_policies.cluster_provided"

Reply:
I see, thanks for the explanation.
The logic is somewhat confusing as the value of `cluster_provided_lb` that may be set inside the `if` scope, is orthogonal (and overrides) the previously set value.

My suggestion to make it a bit clearer to read is this:
```
bool flag;
if (<condition2>) {
  flag = <condition3>;
} else {
  flag = <condition1>;
}
```

But I don't fell strongly about this, so leaving up to you.

Reply:
Then I'll leave it as it is in this PR and may update the logic in the last PR that related to lb extensions.